### ISSUE TITLE
Horovod+PyTorch training is slower than PyTorch DistributedDataParallel

### ISSUE THREAD

**Issue by YouhuiBai** [horovod/horovod#2219]
**Environment:**
1. Framework: (PyTorch)
2. Framework version: 1.3.0a0+1dc2898
3. Horovod version: 0.19.2
4. MPI version: 3.1.2
5. CUDA version: 10.1
6. NCCL version: 2.4.2
7. Python version: 3.6
8. OS and version: CentOS 7.6
9. GCC version: 5.4.0

**Platform:**
4 machines, one 1080Ti GPU per machine, connected with 10Gbps EtherNet

**Checklist:**
1. Did you search issues to find if somebody asked this question before?  Yes
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)?
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)?
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)? Yes

**Your question:**
I trained ImageNet with ResNet50 on above environment and platform using Horovod+PyTorch and PyTorch DDP. Training command of Horovod is as follows:
```sh
$ NCCL_TREE_THRESHOLD=0 NCCL_IB_DISABLE=1 NCCL_SOCKET_IFNAME=NIC horovodrun --network-interface NIC --start-timeout 60 -np 4 -H node1:1,node2:1,node3:1,node4:1 python3 ./pytorch_imagenet_resnet50.py  --train-dir [train dir] --val-dir [val dir] --batch-size 32
```
I tried to enable `autotune` and set multiple nccl streams for `ncclAllreduce`，but the total training throughput of 4 machines is about `530 samples/sec`, meanwhile, the bandwidth usage is about up to `620 MB/sec`.

But when I trianed ResNet50 with PyTorch DDP, the total throughput of 4 machines is up to `700 samples/sec`, and bandwidth usage is about up to `930 MB/sec`, the training command is as follows:
```sh
export NCCL_DEBUG=INFO; export NCCL_TREE_THRESHOLD=0; export NCCL_IB_DISABLE=1; export NCCL_SOCKET_IFNAME=NIC; python pytorch/examples/imagenet/main.py --arch resnet50 -b 32 --gpu 0 --dist-backend 'nccl' --dist-url 'tcp://IP_OF_NODE0:PORT' --world-size 4 --rank [0-3] [data dir]
```
It is confusing that Horovod using `fusion buffer` to increase the bandwidth usage, but PyTorch not, the training speed of Horovod should be higher than PyTorch because ResNet50 contains many little gradients (largest one is 9 MB), however, the training result is far from expectation. Could you explain this phenomenon? Thank you.

**Comment by tgaddair** [comment]
Hey @YouhuiBai, did running with `autotune` suggest different values for the fusion buffer or cycle time?

Also, can you share the output of `horovodrun --check-build` to make sure NCCL is configured correctly?

**Comment by YouhuiBai** [comment]
Hi @tgaddair , thanks for your reply.

For the first question, yes! some of log is as follows:

hierarchical_allreduce,hierarchical_allgather,cache_enabled,cycle_time_ms,tensor_fusion_threshold,score
0,0,0,5,4,371.651
0,0,0,50,32,291.365
0,0,0,25,16,346.455
0,0,0,10,8,381.643
0,0,0,38.616,16.0656,314.045
0,0,0,23.9888,12.0213,341.212
0,0,0,9.99935,8.00034,381.561
0,0,0,49.5129,17.9218,293.789
0,0,0,30.2046,13.1735,332.904
0,0,0,56.6184,6.86872,257.536

For the second question, I can sure that NCCL is configured correctly, because I enable the `NCCL_DEBUG=INFO`, the output of NCCL is correct. The output when enabling `--check-build` is:

Available Frameworks:
    [ ] TensorFlow
    [X] PyTorch
    [ ] MXNet

Available Controllers:
    [X] MPI
    [X] Gloo

Available Tensor Operations:
    [X] NCCL
    [ ] DDL
    [ ] CCL
    [X] MPI
    [X] Gloo

**Comment by YouhuiBai** [comment]
Hi @tgaddair, I also thought that `5 ms` for negotiation cycle is too long, which will increase the latency, especially for ResNet50, therefore I tried to decrease the cycle time to `1 ms` or even `0.1 ms`, the bandwidth usage increasing up to `700 MB/sec`, but still far from `900 MB/sec`.

**Comment by guoshzhao** [comment]
I saw the similar issue when training Resnet101. Horovod+PyTorch training is slower than PyTorch DistributedDataParallel.

Do you have any clues?

**Comment by YouhuiBai** [comment]
@guoshzhao I'm sorry that we don't have any evident clues. I will try to break down the implementation of DistributedDataParallel, maybe find some clues from design or implementation.

**Comment by vdabravolski** [comment]
@YouhuiBai curious if you were able to find a root cause of Horovod slowness compared to DDP.

**Comment by YouhuiBai** [comment]
@vdabravolski Sorry for reply so latter! I was working on other DNN systems like MXNet and TensorFlow recently, didn't take time to understand this problem, but I will because this is interesting. I found that Horovod using negotiation --> fusion --> communication workflow, while DDP using mutiple cuda streams to call `ncclAllReduce()` concurrently, may we can try to think from these difference.

**Comment by philo-zhou** [comment]
@Bert-1208 I have also encountered this problem now.
